Add functions to read VTKs

rthedin opened the pull request:
Add some functions to read VTK and output easier-to-handle format in structured grid with `res` resolution. 

Two options to read a single VTK file:
1. If numpy arrays output:
    `x, y, out = readSingleVTK('full/path/to/vtk/filename.vtk', res=10)`
2. If xarray dataset output:
    `ds = readVTK('full/path/to/timedirs', sliceType='U_zNormal.80.vtk', res=10, t= 25000)`

The anticipated common use is to read many VTKs at once:
    `ds = readVTK('full/path/to/timedirs', sliceType='U_zNormal.80.vtk', res=10, ti= 20000, tf = 25000)`

Additional parameters available. See docstrings.

rthedin committed:
Add functions to read VTKs

rthedin commented:
@ewquon I don't have permissions to request your review.

ewquon commented:
Looks good to me, but is there a reason why this didn't go into `windtools.io.vtk`?

rthedin commented:
The reason is that I did not notice that `io.vtk` existed. Oops. I'll update this PR with a single file dealing with vtks (once Eagle is back online.. sigh). Thanks for the heads up!

ewquon commented on windtools/vtk.py:
```diff
@@ -0,0 +1,339 @@
+    x1d = np.arange(xminbox,xmaxbox+dx,dx)
+    y1d = np.arange(yminbox,ymaxbox+dy,dy)
+    z1d = np.asarray([zlevel])            
+    [x3d,y3d,z3d] = np.meshgrid(x1d, y1d, z1d)
```
Need to add `indexing='ij'`

rthedin replied:
Resolved by commit 804a55790bc95450e35dafc50d017c6c4f26cc0a

ewquon commented on windtools/vtk.py:
```diff
@@ -0,0 +1,339 @@
+    data     = ptdata.GetArray(0)
+
+    # Determine what plane the VTK is in
+    if squash == None:
```
Default behavior should be to return 3D box instead of guessing a squash direction

ewquon commented on windtools/vtk.py:
```diff
@@ -0,0 +1,339 @@
+    data     = ptdata.GetArray(0)
+
+    # Determine what plane the VTK is in
+    if squash == None:
+        # There is a bug with OpenFOAM where a slice in a certain plane gets lots of points also on planes
+        # x=0 or y=0. Here we check the variability of points in all 3 directions and get the direction w/
+        # with less. Example: a slice that is supposed to be at x=1000, will also have of points on x=0
+        nPointsPerDir = [len(np.unique(coords[:,i])) for i in [0,1,2]]
+        dir1 = np.argsort(nPointsPerDir)[2] # the long direction
+        dir2 = np.argsort(nPointsPerDir)[1] # the short direction
+        dirconst = np.argsort(nPointsPerDir)[0]  # 0:xNormal, 1:yNormal, 2:zNormal
```
Default behavior should be to return 3D box instead of guessing a squash direction

rthedin replied:
Resolved by commit efddd1e305291d79e4226766d4aa96e2cd44d076

rthedin committed:
Add `indexing` for consistent meshgrid

rthedin committed:
Reader now works for for generic data varying in any dimension

Reader is now general so that it is not only useful just for slices.
Any VTK containing data varying in all 3 dimensions can now be read. To
resort to the slice reading, the user has to specify the `squash` flag.
Previous behavior was to try to guess in which plane the data lie even
if `squasn` was not provided. Now, without `squash`, it reads the full
3-D field.

rthedin committed:
Combine VTK functions into `windtools.io.vtk`

rthedin commented:
> Looks good to me, but is there a reason why this didn't go into `windtools.io.vtk`?

Resolved by commit https://github.com/NREL/windtools/pull/32/commits/3a6defc02dab8faf33444f51335f09859aab3ecd


@ewquon,
I think this is finally good to be merged. Take a look when you get a chance. Updated instructions are described in the docstrings.

rthedin committed:
Drop variable when squeezing VTK data if `squash!=None`

ewquon commented on windtools/io/vtk.py:
```diff
@@ -10,9 +10,368 @@
+import pandas as pd
+import numpy as np
+
+def readVTK(vtkdir, vtk=None, sliceType=None, dateref=None, ti=None, tf=None, t=None, res=None, squash=None):
```
A minor comment on code design. Maybe `vtkdir` and `vtk=None` can be replaced with `vtkpath`. And then you can check `if os.path.isfile(vtkpath)` then read the single VTK, otherwise `assert os.path.isdir(vtkpath); assert sliceType is not None` for multiples.

rthedin replied:
Resolved by commit https://github.com/NREL/windtools/pull/32/commits/d700ca2562eafaa9e37787d0eb6180cc6676a5b4

ewquon reviewed (commented):
@rthedin FYI

rthedin committed:
Make `readVTK` function more generic

rthedin committed:
Fix typo
